Improve construction tape

ShadowProtocol opened the pull request:
Implements #4479 

# Changes proposed in this pull request:
- Construction tape now behaves like a four way block, connecting to tape in all four directions, with the following exceptions:
    - Tape will remember it is a corner even when isolated (unchanged from previous)
    - Tape will not become a T-junction if it means it will face another T-Junction stem-to-stem. This produces a neat result when building schematics adjacent to one another (makes more sense in the picture)
- Tape is now water-loggable, meaning tape can be (and is) placed in water
    - As a side-effect the tape now sometimes blocks adjacent water sources from flowing in its direction
- Tape placed by the helper (when placing buildings) now replaces flowers
- Tape placed by the helper (when placing buildings) now ignores leaves, replacing them if necessary.

A picture exemplar of all these changes:
![tapesample](https://user-images.githubusercontent.com/30493222/85834572-fafd2580-b7d6-11ea-9910-a6bfc1eb1e29.png)

Review please

ShadowProtocol committed:
Change tape model to act like a FourWayBlock

ShadowProtocol committed:
Add multi-direction logic and default axis

ShadowProtocol committed:
Add dynamic hit box creation and waterlogging

Use the same methods as a FourWayBlock to build the composite hit boxes
Add the ability for tape to be waterlogged
Rename the super class to ...FallingFourWay

ShadowProtocol committed:
Improve tape four way logic

Add condition to remove the stem if there is a T junction opposite another T junction

ShadowProtocol committed:
Rework the Helper

Improve code cleanliness and efficiency in ConstructionTapeHandler
Link to new placement state getters

ShadowProtocol committed:
Add tape corner logic

Ensure tape marked as a corner keeps the correct sides when isolated.

ShadowProtocol committed:
Clean up and fix formatting

Move Four Way shape logic to Abstract tape class
Revert name of abstract falling block class to AbstractBlockMinecoloniesFalling
General formatting and doc fixes

ShadowProtocol committed:
Fix the tape held item model

Add a new model specifically for the inventory like fences have
Adjust the standard model's post top face to not be on a dark patch of wood

ShadowProtocol committed:
Add compatibility

Ensure that tape placed in older versions at least appears, and with the right orientation.

Nightenom commented on src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/ConstructionTapeHelper.java:
```diff
@@ -61,106 +56,77 @@ public static void placeConstructionTape(@NotNull final WorkOrderBuildDecoration
+        final int Y = pos.getY();
+        final int Z = Math.min(corners.getB().getA(), corners.getB().getB());
+        int W = Math.abs(corners.getA().getB() - corners.getA().getA());
+        int H = Math.abs(corners.getB().getB() - corners.getB().getA());
```
use lower case full name, sizeX or similar

Nightenom commented on src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/ConstructionTapeHelper.java:
```diff
@@ -61,106 +56,77 @@ public static void placeConstructionTape(@NotNull final WorkOrderBuildDecoration
-            {
+            
+            if (!world.getBlockState(target).getMaterial().isReplaceable() 
+             && !(world.getBlockState(target).getBlock() instanceof LeavesBlock)
```
leaves block only when not persistent, or I would say no to all

ShadowProtocol replied:
Not quite sure what you are getting at. The point of that line is so that overhanging trees are ignored. What do you mean?

Raycoms replied:
There is a property on leaves if they are placed manually by the player or builder. Those should not be replaced.

Nightenom replied:
I would say no to all, unless you can find air block under a row of leaves only

ShadowProtocol replied:
Now does a scan for air (or something properly replaceable) below.

Nightenom reviewed (commented):
Formatting in different places
Otherwise good job 👍

ShadowProtocol committed:
Avoid replacement of leaves

Use a forward scan to check if there is something replaceable below leaves instead
Fix some capitalization

Raycoms commented on src/api/java/com/minecolonies/api/blocks/decorative/AbstractBlockMinecoloniesConstructionTape.java:
```diff
@@ -1,107 +1,125 @@
-            if (tempMeta < 0 || tempMeta >= META_LOOKUP.length)
-            {
-                tempMeta = 0;
+            if (p_223007_0_.get(SOUTH)) {
```
just the formatting is off here yet.

Raycoms commented on src/api/java/com/minecolonies/api/blocks/decorative/AbstractBlockMinecoloniesConstructionTape.java:
```diff
@@ -1,107 +1,125 @@
-            return this.unlocalizedName;
+        // Combine the arm voxel shapes with the main node for all combinations
+        for(int i = 0; i < 16; ++i) {
+            avoxelshape[i] = VoxelShapes.or(node, avoxelshape[i]);
```
here too

ShadowProtocol committed:
Fix formatting and naming

Raycoms commented on src/api/java/com/minecolonies/api/blocks/decorative/AbstractBlockMinecoloniesConstructionTape.java:
```diff
@@ -1,107 +1,121 @@
-        {
-            this(metaIn, nameIn, nameIn, mapColorIn);
-        }
+    private static int getMask(Direction facing) {
```
format

ShadowProtocol replied:
I'll activate the IDE file formatter with the right settings now... although I don't know how I missed that many. I did try.

ShadowProtocol committed:
Fix formatting
